File: node-express/nodejs/src/services/data-service.js
const { Client } = require('pg');
const { DataError } = require('../errors/data-error');

const client = new Client({
  host: process.env.DATABASE_HOST,
  port: process.env.DATABASE_PORT,
  user: process.env.DATABASE_USER,
  password: process.env.DATABASE_PASSWORD,
  database: process.env.DATABASE_NAME,
});

const SEPARATOR = ', ';
const PARAM_PREFIX = '$'

const initialize = async () => {
  try {
    console.log('Opening database connection');
    await client.connect();
  } catch (err) {
    console.error('Failed to open database connection: ', err);
  }
}

const finalize = async () => {
  try {
    console.log('Closing datbase connection');
    await client.end();
    return true;
  } catch (err) {
    console.error('Failed to close database connection: ', err);
    return false;
  }
}

const buildSelectClause = (columns) => {
  var str = 'SELECT ';
  if (columns == null || columns.length == 0) {
    str += '*';
  } else {
    str += columns.join(SEPARATOR);
  }

  return str;
}

const buildWhereClause = (filters, i = 1) => {
  if (filters == null || Object.keys(filters).length == 0) {
    return '';
  }

  var str = ' WHERE ';
  let pairs = [];
  Object.keys(filters).forEach(key => {
    pairs.push(`${key} = ${PARAM_PREFIX}${i}`);
    i++;
  });
  str += pairs.join(' AND ');

  return str;
}

const buildValuesClause = (values, i = 1) => {
  if (values == null || Object.keys(values).length == 0) {
    return '';
  }

  var str = ' VALUES ';
  let pairs = [];
  Object.keys(values).forEach(key => {
    pairs.push(`${PARAM_PREFIX}${i}`);
    i++;
  });
  str += `(${pairs.join(SEPARATOR)})`;

  return str;
}

const buildReturningClause = (columns) => {
  if (columns == null || columns.length == 0) {
    return '';
  }

  var str = ' RETURNING ';
  str += columns.join(SEPARATOR);

  return str;
}

const buildSetClause = (values, i = 1) => {
  if (values == null || Object.keys(values).length == 0) {
    return '';
  }

  var str = ' SET ';
  let pairs = [];
  Object.keys(values).forEach(key => {
    pairs.push(`${key} = ${PARAM_PREFIX}${i}`);
    i++;
  });
  str += pairs.join(SEPARATOR);

  return str;
}

const querySelect = async (table, count, offset = null, columns = null, filters = null) => {
  var sql = buildSelectClause(columns);
  sql += ` FROM ${table}`;
  sql += buildWhereClause(filters);
  sql += ` LIMIT ${count}`;
  if (offset != null) {
    sql += ` OFFSET ${offset}`;
  }

  try {
    const res = await client.query(sql, filters != null
      ? Object.values(filters) : null);
    return res.rows;
  } catch (err) {
    console.error('Failed to execute select query: ' + err);
    throw new DataError('Database operation failed', { cause: err });
  }
}

const querySelectFirst = async (table, columns = null, filters = null) => {
  const rows = await querySelect(table, 1, null, columns, filters);
  return rows.length > 0 ? rows[0] : null;
}

const queryInsert = async (table, values, retColumns = null) => {
  if (values == null || Object.keys(values).length == 0) {
    throw new DataError('Values cannot be null or empty');
  }

  var sql = `INSERT INTO ${table} (${Object.keys(values).join(SEPARATOR)})`;
  sql += buildValuesClause(values);
  sql += buildReturningClause(retColumns);

  try {
    const res = await client.query(sql, Object.values(values));
    if (retColumns != null && res.rows.length > 0) {
      return res.rows[0];
    } else if (res.rowCount > 0) {
      return res.rowCount;
    }
  } catch (err) {
    console.error('Failed to execute insert query: ' + err);
    throw new DataError('Database operation failed', { cause: err });
  }

  throw new DataError('Database operation failed');
}

const queryUpdate = async (table, values, filters = null, retColumns = null) => {
  if (values == null || Object.keys(values).length == 0) {
    throw new DataError('Values cannot be null or empty');
  }

  var sql = `UPDATE ${table}`;
  sql += buildSetClause(values);
  sql += buildWhereClause(filters, Object.keys(values).length + 1);
  sql += buildReturningClause(retColumns);

  try {
    const res = await client.query(sql, filters != null
      ? Object.values(values).concat(Object.values(filters)) : Object.values(values));
    if (retColumns != null) {
      return res.rows.length > 0 ? res.rows[0] : null;
    } else {
      return res.rowCount;
    }
  } catch (err) {
    console.error('Failed to execute update query: ' + err);
    throw new DataError('Database operation failed', { cause: err });
  }
}

const queryDelete = async (table, filters = null) => {
  var sql = `DELETE FROM ${table}`;
  sql += buildWhereClause(filters);

  try {
    const res = await client.query(sql, filters != null
      ? Object.values(filters) : null);
    return res.rowCount;
  } catch (err) {
    console.error('Failed to execute delete query: ' + err);
    throw new DataError('Database operation failed', { cause: err });
  }
}

initialize();

module.exports = {
  finalize,
  querySelect,
  querySelectFirst,
  queryInsert,
  queryUpdate,
  queryDelete
};

File: csharp-dotnet/dotnet/src/Services/DataService.cs
using System.Data;
using System.Text;
using MySql.Data.MySqlClient;

namespace WebApi.Services
{
    public class DataService : IDisposable
    {
        private const string SEPARATOR = ", ";
        private const string PARAM_PREFIX_FILTER = "@f_";
        private const string PARAM_PREFIX_VALUE = "@v_";

        private readonly ILogger<DataService> _logger;
        private MySqlConnection _connection;
        private bool _disposed;

        public DataService(ILogger<DataService> logger)
        {
            _logger = logger;
            try
            {
                _logger.LogInformation("Openning database connection");
                var connectionString = string.Format(
                    "server={0};port={1};uid={2};pwd={3};database={4}",
                    Environment.GetEnvironmentVariable("DATABASE_HOST"),
                    Environment.GetEnvironmentVariable("DATABASE_PORT"),
                    Environment.GetEnvironmentVariable("DATABASE_USERNAME"),
                    Environment.GetEnvironmentVariable("DATABASE_PASSWORD"),
                    Environment.GetEnvironmentVariable("DATABASE_NAME")
                );
                _connection = new MySqlConnection(connectionString);
                _connection.Open();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to open database connection");
                _connection = null;
            }
        }

        #region Private Helper Methods
        private static string BuildSelectClause(string[] columns)
        {
            var str = "SELECT ";
            if (columns == null || columns.Length == 0)
            {
                str += "*";
            }
            else
            {
                str += string.Join(SEPARATOR, columns);
            }
            return str;
        }

        private static string BuildWhereClause(IDictionary<string, object> filters)
        {
            if (filters == null || filters.Count == 0) return "";

            var str = " WHERE ";
            var pairs = new List<string>();
            foreach (var key in filters.Keys)
            {
                pairs.Add($"{key} = {PARAM_PREFIX_FILTER}{key}");
            }

            str += string.Join(" AND ", pairs);
            return str;
        }

        private static string BuildValuesClause(IDictionary<string, object> values)
        {
            if (values == null || values.Count == 0) return "";

            var str = " VALUES ";
            var pairs = new List<string>();
            foreach (var key in values.Keys)
            {
                pairs.Add($"{PARAM_PREFIX_VALUE}{key}");
            }

            str += $"({string.Join(SEPARATOR, pairs)})";
            return str;
        }

        private static string BuildSetClause(IDictionary<string, object> values)
        {
            if (values == null || values.Count == 0) return "";

            var str = " SET ";
            var pairs = new List<string>();
            foreach (var key in values.Keys)
            {
                pairs.Add($"{key} = {PARAM_PREFIX_VALUE}{key}");
            }

            str += string.Join(SEPARATOR, pairs);
            return str;
        }

        private static void AddParameterValues(
            MySqlParameterCollection parameters,
            IDictionary<string, object> values = null,
            IDictionary<string, object> filters = null)
        {
            if (values != null)
            {
                foreach (var value in values)
                {
                    parameters.AddWithValue($"{PARAM_PREFIX_VALUE}{value.Key}", value.Value);
                }
            }

            if (filters != null)
            {
                foreach (var filter in filters)
                {
                    parameters.AddWithValue($"{PARAM_PREFIX_FILTER}{filter.Key}", filter.Value);
                }
            }
        }
        #endregion

        #region Public Query Methods
        public async Task<IEnumerable<T>> QuerySelectAsync<T>(
            IDataCallback callback,
            string table,
            int count,
            int? offset = null,
            string[] columns = null,
            IDictionary<string, object> filters = null)
        {
            ArgumentNullException.ThrowIfNull(callback, nameof(callback));
            if (_connection == null)
                throw new InvalidOperationException("Database connection is null");

            try
            {
                var sqlBuilder = new StringBuilder();
                sqlBuilder.Append(BuildSelectClause(columns));
                sqlBuilder.Append($" FROM {table}");
                sqlBuilder.Append(BuildWhereClause(filters));
                sqlBuilder.Append($" LIMIT {count}");
                if (offset != null)
                {
                    sqlBuilder.Append($" OFFSET {offset}");
                }

                var sqlCommand = new MySqlCommand(sqlBuilder.ToString(), _connection);
                AddParameterValues(sqlCommand.Parameters, null, filters);
                using var reader = await sqlCommand.ExecuteReaderAsync(CommandBehavior.Default);
                return callback.ParseResultSet<T>(reader);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to execute select query");
                throw new DataException("Database operation failed", ex);
            }
        }

        public async Task<T> QuerySelectFirstAsync<T>(
            IDataCallback callback,
            string table,
            string[] columns = null,
            IDictionary<string, object> filters = null)
        {
            var results = await QuerySelectAsync<T>(callback, table, 1, null, columns, filters);
            return results != null ? results.FirstOrDefault() : default;
        }

        public async Task<long> QueryInsertAsync(
            string table,
            IDictionary<string, object> values)
        {
            if (values == null || values.Count == 0)
                throw new ArgumentException("Value cannot be null or empty", nameof(values));
            if (_connection == null)
                throw new InvalidOperationException("Database connection is null");

            try
            {
                var sqlBuilder = new StringBuilder();
                sqlBuilder.Append($"INSERT INTO {table} ({string.Join(SEPARATOR, values.Keys)})");
                sqlBuilder.Append(BuildValuesClause(values));

                var sqlCommand = new MySqlCommand(sqlBuilder.ToString(), _connection);
                AddParameterValues(sqlCommand.Parameters, values, null);
                return await sqlCommand.ExecuteNonQueryAsync() > 0
                    ? sqlCommand.LastInsertedId
                    : throw new DataException("Database operation failed");
            }
            catch (DataException)
            {
                throw;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to execute insert query");
                throw new DataException("Database operation failed", ex);
            }
        }

        public async Task<int> QueryUpdateAsync(
            string table,
            IDictionary<string, object> values,
            IDictionary<string, object> filters = null)
        {
            if (values == null || values.Count == 0)
                throw new ArgumentException("Value cannot be null or empty", nameof(values));
            if (_connection == null)
                throw new InvalidOperationException("Database connection is null");

            try
            {
                var sqlBuilder = new StringBuilder();
                sqlBuilder.Append($"UPDATE {table}");
                sqlBuilder.Append(BuildSetClause(values));
                sqlBuilder.Append(BuildWhereClause(filters));

                var sqlCommand = new MySqlCommand(sqlBuilder.ToString(), _connection);
                AddParameterValues(sqlCommand.Parameters, values, filters);
                return await sqlCommand.ExecuteNonQueryAsync();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to execute update query");
                throw new DataException("Database operation failed", ex);
            }
        }

        public async Task<int> QueryDeleteAsync(
            string table,
            IDictionary<string, object> filters = null)
        {
            if (_connection == null)
                throw new InvalidOperationException("Database connection is null");

            try
            {
                var sqlBuilder = new StringBuilder();
                sqlBuilder.Append($"DELETE FROM {table}");
                sqlBuilder.Append(BuildWhereClause(filters));

                var sqlCommand = new MySqlCommand(sqlBuilder.ToString(), _connection);
                AddParameterValues(sqlCommand.Parameters, null, filters);
                return await sqlCommand.ExecuteNonQueryAsync();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to execute delete query");
                throw new DataException("Database operation failed", ex);
            }
        }
        #endregion

        protected virtual void Dispose(bool disposing)
        {
            if (_disposed) return;

            if (disposing)
            {
                _connection?.Close();
                _connection = null;
            }

            _disposed = true;
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
    }
}
